metagraph/src/graph/representation/succinct/boss_chunk_construct.cpp

```diff
@@ -673,114 +710,225 @@ void recover_dummy_nodes_disk(KmerCollector &kmer_collector,
+                        const std::vector<std::string> &dummy_sink_names,
+                        std::filesystem::path dir,
+                        size_t k,
+                        bool both_strands_mode,
+                        uint8_t bits_per_count,
+                        std::filesystem::path swap_dir,
+                        size_t num_threads) {
+    assert(real_names.size() == dummy_sink_names.size());
+    assert(real_names.size() + 1 == dummy_source_names.front().size());
+    assert(dummy_source_names.size() == k);
+
+    if (k <= 1) {
+        // For k <= 1 BOSS chunks can't be built independently.
+        // Concatenate blocks and build a full BOSS table.
+        logger->trace("Concatenating blocks of real k-mers ({} -> 1)...", real_names.size());
+        std::string real_name = dir/"real_kmers";
+        common::concat(real_names, real_name);
+
+        logger->trace("Concatenating blocks of dummy sink k-mers ({} -> 1)...",
+                      dummy_sink_names.size());
+        std::vector<std::string> dummy_names { dir/"dummy_sink" };
+        common::concat(dummy_sink_names, dummy_names.back());
+
+        for (size_t i = 0; i < dummy_source_names.size(); ++i) {
+            logger->trace("Concatenating blocks of dummy-{} source k-mers ({} -> 1)...",
+                          i + 1, dummy_source_names[i].size());
+            dummy_names.push_back(dir/fmt::format("dummy_source_{}", i));
+            common::concat(dummy_source_names[i], dummy_names.back());
+        }
+
+        return build_boss_chunk<T>(true, real_name, dummy_names,
+                                   k, both_strands_mode, bits_per_count, swap_dir);
+    }
+
+    // For k >= 2, BOSS chunks can be built independently, in parallel.
+
+    // there are real k-mers that end with $, so we do the following trick
+    const std::string empty_real_name = real_names.front() + "_empty";
+    Encoder<get_int_t<T>> empty_real(empty_real_name, ENCODER_BUFFER_SIZE);
+    empty_real.finish();
+
+    // the first chunk has at most alphabet_size + 1 k-mers (where F = $)
+    std::vector<std::string> dummy_names;
+    for (size_t i = 0; i < dummy_source_names.size(); ++i) {
+        dummy_names.push_back(dummy_source_names[i][0]);
+    }
+    BOSS::Chunk *chunk = build_boss_chunk<T>(true, empty_real_name, dummy_names,
+                                             k, both_strands_mode, bits_per_count, swap_dir);
+    logger->trace("Chunk for ..$. done");
+    // construct all other chunks in parallel
+    #pragma omp parallel for ordered num_threads(num_threads) schedule(dynamic)
+    for (size_t F = 0; F < real_names.size(); ++F) {
+        std::vector<std::string> dummy_names { dummy_sink_names[F] };
+        for (size_t i = 0; i < dummy_source_names.size(); ++i) {
+            dummy_names.push_back(dummy_source_names[i][F + 1]);
+        }
+        BOSS::Chunk *next = build_boss_chunk<T>(false, real_names[F], dummy_names,
```

Comment:
at some point, we should have a TODO to make these return smart pointers instead